README.txt
DEBUG CHEATSHEET
(GERMAN VERSION SEE BELOW)

Since the a21glossary is partially tricky to adjust, this part delivers useful bavioral feedback.

- You need to be logged into the Typo3-Backend, or the debug parameters wont work
- Typo3 Page Caching is disabled when using one of these options, so feel free to experiment.

How to use:
Simply add the params to your URL (POST is also supported if you want).
Alternately you can set then using typoscript (e.g. “config.tx_a21glossary.debug.info = 1”)


Very important:

	&tx_a21glossary[noncaching]=1	marks non-caching areas (if any), interesting to find potential bottlenecks
	-> red dashed for internal Typo3-Elements
	-> red dotted for external PHP-Scripts

	&tx_a21glossary[markers]=1		highlights the content being replaced by the glossary (green), highlights excluded parts/tags (red)

Important:

	&tx_a21glossary[highlight]=1	highlights the replaced glossar items in green for better recognition
	&tx_a21glossary[info]=1			shows some glossary stats below the content, eg. the runtime
	&tx_a21glossary[conf]=1			show glossary config below content
	&tx_a21glossary[trail]=1		who triggered the a21glossary (trail of function calls)

Less important:

	&tx_a21glossary[disable]=1		disables the glossary functionality
	&tx_a21glossary[query]=1		debug query to find glossary elements, one for each pid
	&tx_a21glossary[result]=1		debug all found glossary elements
	&tx_a21glossary[regexp]=1		show regular expression config below content
	&tx_a21glossary[keep]=1			keep markers in source after processing

Unimportant:

	&tx_a21glossary[input]=1		show raw input
	&tx_a21glossary[output]=1		show raw output
	&tx_a21glossary[demo]=1			add some demo-content and demo word ("etc.") into content




##############################
# in Deutsch
##############################

- Damit die Paramter anerkannt werden, muss man im Backend eingeloggt sein
- Auf der entsprechenden Frontend-Seite muss das Admin-Panel erscheinen, dort sicherheitshalber das Nicht-Cachen erzwingen
- dann die Parameter an die URL anfügen


Sehr wichtig:

	&tx_a21glossary[noncaching]=1	HEBT NICHT CACHENDE SEITENINHALTE HERVOR (SOFERN VORHANDEN)
	-> rot gestrichelt (dashed) for interne Typo3-Elemente
	-> rot Gepunktet (dotted) für externe PHP-Skripte

	&tx_a21glossary[markers]=1		Hervorhebung der aktiven Glossarbereiche in Grün, ausgeschlossene Bereiche in Rot

Wichtig:

	&tx_a21glossary[highlight]=1	Stärkere Hervorhebung der Glossarelemente (Neongrün)
	&tx_a21glossary[info]=1 		zeigt einige Statistiken (z.B. Laufzeit) an
	&tx_a21glossary[conf]=1			Zeigt die Glossarkonfiguration an (z.B. pids, excludeTags ...)
	&tx_a21glossary[trail]=1		Wer hat das Glossar getriggert (Trail der Funktionsaufrufe)

Weniger wichtig:

	&tx_a21glossary[disable]=1		deaktiviert das Glossar
	&tx_a21glossary[query]=1		SQL-Query, um Glossarelemente aufzufinden
	&tx_a21glossary[result]=1		SQL-Ergebnis der gefundenen Elemente
	&tx_a21glossary[regexp]=1		erstellte Regular Expressions
	&tx_a21glossary[keep]=1			keep markers in source after processing

Nicht Empfohlen:

	&tx_a21glossary[input]=1		content vor der verarbeitung
	&tx_a21glossary[output]=1		content nach der Verarbeitung
	&tx_a21glossary[demo]=1			Seiteninhalt mit Democonent und Glossar um Demobegriff ("etc.") erweitern



ChangeLog


14-04-08 (6.2.0) Ronny Vorpahl <vorpahl@artplan21.de>

	* [API] added Typo3 6.2 compatibility
		* removed debug statements and all related stuff (see typo3 coding guidelines)
		* removed xclass definitions
		* added ext_autoload.php
		* updated ext_emconf.php


14-04-08 (0.9.5) Ronny Vorpahl <vorpahl@artplan21.de>

	* [BUGFIX] #40554: Link generation fails for translated acronyms
		* removed spaces from dataWrap
		* Thanks to Philipp Kitzberger for the solution


11-12-08 (0.9.4) Ronny Vorpahl <vorpahl@artplan21.de>

	* [BUGFIX] #31929: FE-plugin ignores storage page configuration
		* Thanks to Michael Schreiner for the solution
	* [BUGFIX] #31930: Debug options are activated on glossary detail page (if logged into Typo3-Backend)
		* thanks to Michael Schreiner for the hint
	* Synced both typoscript configurations (old+new) as far as possible
		* they differ only in css/wraps/list-styling - no behavioral differences

	* TER upload


11-11-21 (0.9.3) Oliver Hader <oliver.hader@typo3.org>

	* added a second (more accessible) typoscript-configuration in folder "Configuration"
		* glossary elements are wrapped with correct CSS classes with extension prefixes
	* consistent usage of LL-XML in TCA
	* moved Tca.php from root folder to "Configuration"


11-11-04 (0.9.2) Oliver Hader <oliver.hader@typo3.org>

	* [BUGFIX] #31571: Wraps produce incorrect nested paragraph tags


11-11-03 (0.9.1) Ronny Vorpahl <vorpahl@artplan21.de>

	* Thanks to Xaver Maierhofer: changed <p class=back" -> <p class="back" in constants.txt
	* minor label changes


11-11-01 (0.9.0) Ronny Vorpahl <vorpahl@artplan21.de>

	* committed to TER

11-11-01 (0.8.4) Ronny Vorpahl <vorpahl@artplan21.de>

	* fixed sugesstion #31006:  changed Label of tx_a21glossary_main.link as suggested in

	* committing Bugfix/Maintenance Release from Oliver Hader - Thank you very much by the way
		53539	2011-10-28 19:36	Oliver Hader	[TASK] Update ext_emconf.php
		53538	2011-10-28 19:33	Oliver Hader	[BUGFIX] #31388: Applying exclude tags is plain wrong
		53535	2011-10-28 18:51	Oliver Hader	[TASK] Add missing locallang.xml file
		53534	2011-10-28 18:42	Oliver Hader	[TASK] Update ext_emconf.php
		53533	2011-10-28 18:39	Oliver Hader	[TASK] Add missing comment constant for anchor solution
		53532	2011-10-28 18:37	Oliver Hader	[FEATURE] #31007: Integrate labels of types shown in frontend to locallang files
		53531	2011-10-28 18:12	Oliver Hader	[BUGFIX] #31009: Paragraphs in description are not rendered in frontend
		53530	2011-10-28 18:08	Oliver Hader	[BUGFIX] Constant storagePage can be a list
		53529	2011-10-28 18:03	Oliver Hader	[BUGFIX] #31386: Wrong RTE configuration in TCA
		53527	2011-10-28 17:44	Oliver Hader	[BUG] #31383: PHP warning if processed before caching

11-10-05 (0.8.3) Ronny Vorpahl <vorpahl@artplan21.de>

	* Moved three basic settings into extension config, still overridable by typoscript
		* glossaryWHEN
			* before Caching: faster, but non-caching content won't be affected (recommended)
			* after Caching: slower, but whole even uncached content will be affected
		* glossaryWHERE
			* whole content (old style)
			* use Typo3Search-Tags (recommended if using idexed_search)
			* i will set my own inclusions/exclusions using a21glossary-markers (see manual, they work in all modes)
		* excludeTags
	* updated documentation and README.txt (Debug Cheatsheet)
	* added clearCacheOnLoad = 1


11-10-04 (0.8.2) Ronny Vorpahl <vorpahl@artplan21.de>

	* major cleanup, partially rewrite of a21glossary algorithm


11-09-30 (0.8.1) Ronny Vorpahl <vorpahl@artplan21.de>

	* preg_quote and preg_replace is reguired now - no more workarounds
	* minor cleanup


11-08-29 (0.8.0) Benjamin Mack <benni@typo3.org>

	* Typo3 4.5 compatibility enhancement
	* code cleanup according to Typo3 Coding guidelines

	* behavioral change
		glossary starts without explicit inclusion tags: $this->splitAndReplace($document, 1);
		glossary will only start if inclusion settings are met: $this->splitAndReplace($document);


10-11-24 (0.7.3) Ronny Vorpahl <vorpahl@artplan21.de>

	* removed line break error


10-11-24 (0.7.2) Ronny Vorpahl <vorpahl@artplan21.de>

	* maintenance
		* removed ereg_replace, PREG_REPLACE IS NOW MANDATORY ('about time)
		* translated README.txt into german
		* improved debug:noncaching, regexp
		* debug marksers implies now noncaching
		* removed last window line breaks from several files


10-11-23 (0.7.1) Ronny Vorpahl <vorpahl@artplan21.de>

	* fixed double entry in excludeTags (appeared in 0.7.0), glossary now with now with double entry check

	* maintenance: debug commands for frontend usage
		* added debugQuery functions, use  &tx_a21glossary[query]=1
		* added debugResult functions, use  &tx_a21glossary[result]=1
		* added keepMarkers functions, use  &tx_a21glossary[keep]=1
		* show included and excluded page parts &tx_a21glossary[markers]=1
		* highlight replaced words &tx_a21glossary[highlight]=1
		* other commands (use like "highlight"):
			info, conf, noncaching,regexp,output,input

	* maintenance: other tasks
		* minor cleanup of comments
		* inval()'d many params
		* added README.txt with debug hints
		* TODO update manual, see README.TXT for new debug commands


10-10-05 (0.7.0) Ronny Vorpahl <vorpahl@artplan21.de>

	* MAJOR BEHAVIORAL CHANGE:

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "a21glossary".
 *
 * Auto generated 08-04-2014 13:08
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'A21 Glossary',
	'description' => 'A21 Glossary - automatical conversion of all abbreviations and acronyms in the special tags for accessibility issues',
	'category' => 'fe',
	'shy' => 0,
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'internal' => 0,
	'uploadfolder' => 0,
	'createDirs' => 'uploads/tx_a21glossary/rte/',
	'modify_tables' => '',
	'clearCacheOnLoad' => 1,
	'lockType' => '',
	'author' => 'Ronny Vorpahl',
	'author_email' => 'info@artplan21.de',
	'author_company' => 'Artplan21 GmbH. - www.artplan21.de',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'version' => '6.2.0',
	'_md5_values_when_last_written' => 'a:18:{s:9:"ChangeLog";s:4:"d6a8";s:24:"class.tx_a21glossary.php";s:4:"16f1";s:16:"ext_autoload.php";s:4:"5489";s:21:"ext_conf_template.txt";s:4:"76ef";s:12:"ext_icon.gif";s:4:"dff2";s:17:"ext_localconf.php";s:4:"b780";s:15:"ext_php_api.dat";s:4:"a346";s:14:"ext_tables.php";s:4:"382e";s:14:"ext_tables.sql";s:4:"f3ae";s:28:"icon_tx_a21glossary_main.gif";s:4:"a4ed";s:13:"locallang.xml";s:4:"efca";s:16:"locallang_db.xml";s:4:"4280";s:21:"Configuration/Tca.php";s:4:"36e1";s:52:"Configuration/TypoScript/Accessibility/constants.txt";s:4:"8dbf";s:48:"Configuration/TypoScript/Accessibility/setup.txt";s:4:"9a24";s:14:"doc/manual.sxw";s:4:"f0a4";s:20:"static/constants.txt";s:4:"26fe";s:16:"static/setup.txt";s:4:"df90";}',
	'constraints' => array(
		'depends' => array(
			'php' => '5.3.0-0.0.0',
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
			'indexed_search' => '0.0.1-',
		),
	),
	'suggests' => array(
	),
);

?>

ext_tables.php
<?php

if (!defined('TYPO3_MODE')) {
	die ('Access denied.');
}

t3lib_extMgm::addStaticFile($_EXTKEY, 'static/', 'A21 Glossary - Default Output');
t3lib_extMgm::addStaticFile($_EXTKEY, 'Configuration/TypoScript/Accessibility/', 'A21 Glossary - Accessible Output');

t3lib_extMgm::addPlugin(array('LLL:EXT:a21glossary/locallang_db.xml:tx_a21glossary_main', 'a21glossary'));
t3lib_extMgm::allowTableOnStandardPages('tx_a21glossary_main');
t3lib_extMgm::addToInsertRecords('tx_a21glossary_main');

$TCA['tx_a21glossary_main'] = array(
	'ctrl' => array(
		'title' => 'LLL:EXT:a21glossary/locallang_db.xml:tx_a21glossary_main',
		'label' => 'short',
		'tstamp' => 'tstamp',
		'crdate' => 'crdate',
		'cruser_id' => 'cruser_id',
		'versioning' => '1',
		'languageField' => 'sys_language_uid',
		'transOrigPointerField' => 'l18n_parent',
		'transOrigDiffSourceField' => 'l18n_diffsource',
		'default_sortby' => 'ORDER BY short,uid',
		'delete' => 'deleted',
		'enablecolumns' => array(
			'disabled' => 'hidden',
			'starttime' => 'starttime',
			'endtime' => 'endtime',
			'fe_group' => 'fe_group'
		),
		'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY) . 'Configuration/Tca.php',
		'iconfile' => t3lib_extMgm::extRelPath($_EXTKEY) . 'icon_tx_a21glossary_main.gif',
	),
	'feInterface' => array(
		'fe_admin_fieldList' => 'sys_language_uid, l18n_parent, l18n_diffsource, hidden, starttime, endtime, fe_group, short, shortcut, longversion, shorttype, language, description, link, exclude, force_linking, force_case, force_preservecase, force_regexp, force_global',
	)
);

?>

ext_localconf.php
<?php
if (!defined ("TYPO3_MODE")) 	die ("Access denied.");
t3lib_extMgm::addPageTSConfig('

	# ***************************************************************************************
	# CONFIGURATION of RTE in table "tx_a21glossary_main", field "description"
	# ***************************************************************************************
RTE.config.tx_a21glossary_main.description {
  hidePStyleItems = H1, H4, H5, H6
  proc.exitHTMLparser_db=1
  proc.exitHTMLparser_db {
    keepNonMatchedTags=1
    tags.font.allowedAttribs= color
    tags.font.rmTagIfNoAttrib = 1
    tags.font.nesting = global
  }
}
');
$TYPO3_CONF_VARS['SC_OPTIONS']['tslib/class.tslib_fe.php']['contentPostProc-all']['tx_a21glossary'] = 'EXT:a21glossary/class.tx_a21glossary.php:&tx_a21glossary->convertGlossaryWords';
?>

class.tx_a21glossary.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2011 Ronny Vorpahl (vorpahl@artplan21.de)
*  (c) 2003 Andreas Schwarzkopf (schwarzkopf@artplan21 de)
*  All rights reserved
*
*  This script is part of the Typo3 project. The Typo3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * a21glossary: all found words in content wich correspond with the glossary entries
 * will be enriched with special markup and/or with links to the glossary details page
 *
 * @author	Ronny Vorpahl <vorpahl@artplan21.de>
 */

class tx_a21glossary {

	protected $search = array();
	protected $searchGlobal = array();
	protected $count = array();
	protected $used = array();
	protected $searchMarkers = array();
	protected $replaceMarkers = array();
	protected $searchMarkers2 = array();
	protected $replaceMarkers2 = array();

	/**
	 * function call to apply to the totally rendered page (with non-caching
	 * objects) hook the alternative is the userfunc below
	 *
	 * @param string $content the full HTML content to output as object
	 * @param tslib_fe $pObj the parent object, in this case the TSFE global object
	 * @return void
	 */
	public function processHook(&$content, $pObj) {

		$conf = $GLOBALS['TSFE']->config['config']['tx_a21glossary.'];
		$content['pObj']->content = $this->main($content['pObj']->content, $conf);
	}

	/**
	 * function call to do the processing via a user function
	 * the alternative is to apply it to the whole HTML content (see above)
	 *
	 * @param string $content the HTML content to output
	 * @param array $conf the configuration array of the user function
	 * @return string the modified HTML content to output
	 */
	public function processUserFunc($content, $conf) {

		if (empty($conf)) {
			$conf = $GLOBALS['TSFE']->config['config']['tx_a21glossary.'];
		}
		return $this->main($content, $conf);
	}


	/**
	 * this is the actual main function that replaces the glossary
	 * words with the explanations
	 *
	 * @param string $content the content that should be parsed
	 * @param array $conf the configuration for the parsing
	 * @return string the modified content
	 */
	protected function main($content, array $conf = NULL) {

		GLOBAL $TSFE;

		$this->time_start = microtime(true);

		// merge with extconf, $conf overrules
		$extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['a21glossary']);
		if(count($extConf)) {
			$conf = array_merge($extConf, (array) $conf);
		}

		// return right now if the wrong page type was chosen
		$typeList = $TSFE->config['typeList'];
		$typeList = strlen($typeList) ? $typeList : '0';
		$typeList = @explode(',', $typeList);

		if (!in_array(intval(t3lib_div::_GP('type')), $typeList)) {
			return $content;
		}

		// load the whole configuration
		$id = $TSFE->id;
		$pageLang = $TSFE->config['config']['language'] ? $TSFE->config['config']['language'] : $TSFE->config['config']['htmlTag_langKey'];
		$renderCharset = $TSFE->renderCharset;

		// extract and escape get-vars
		$this->piVars = t3lib_div::_GP('tx_a21glossary');
		if(count($this->piVars)) {
			t3lib_div::addSlashesOnArray($this->piVars);
		}

		// for the stats
		for ($i = -1; $i<=15; $i++) {
			$this->depths[$i] = 0;
		}

		if (isset($conf['noglossary']) || $this->piVars['disable']) {
			return $content;
		}

		if (t3lib_div::inList($conf['excludePages'], $id)) {
			if ($conf['excludePages.'][$id]) {
				// disable glossary only for certain glossary types
				$conf['excludeTypes'] .= ',' . $conf['excludePages.'][$id];
			} else {
				// disable glossary completly on current page: stop glossary rendering immediately
				return $content;
			}
		}

		$items = $this->fetchGlossaryItems($conf['pidList']);

		if (!$this->count['used']) {
			return $content;
		} else {

			$cObj = t3lib_div::makeInstance('tslib_cObj');

			// sort entries from z-a to replace more special words with the same beginnng first eg. aminoacid before amino
			krsort($items);

			// move entries with differing shortcut to end of array to prevent double replacement
			foreach ($items as $key => $item) {
				if (strlen($item['shortcut']) && ($item['shortcut'] != $item['short'])) {
					unset($items[$key]);
					$items[$key] = $item;
				}
			}

			// prepare items
			foreach ($items as $item) {

				if (!t3lib_div::inList($conf['excludeTypes'], $item['shorttype'])) {

					$cObj->data = $item;

					// set item language
					if ($item['language'] && $pageLanguage != $item['language']) {
						$lang  = (intval($conf['noLang']) ? '' : (' lang="'.$item['language'].'"'))
							. (intval($conf['xmlLang']) ? (' xml:lang="'.$item['language'].'"') : '');
					} else {
						$lang = '';
					}

					// set item type
					$element	= trim(htmlspecialchars(strip_tags($item['shorttype']), ENT_QUOTES, $renderCharset));
					$titleText	= trim(htmlspecialchars(strip_tags($item['longversion']), ENT_QUOTES, $renderCharset));
					$title		= $item['longversion'] ? (' title="' . $titleText . '"') : '';

					// those can be retrieved later with stdwrap
					$TSFE->register['lang'] = $lang;
					$TSFE->register['title'] = $title;

					// decide replacement linking
					if ($item['force_linking']) {
						$generateLink = ($item['force_linking'] == 1) ? 1 : 0;

					} elseif (t3lib_div::inList($conf['linkToGlossary'], $item['shorttype']) && count($conf['typolink.'])) {
						$generateLink = 1;

						if (strlen($conf['linkOnlyIfNotEmpty'])) {
							$linkOnlyIfNotEmpty = t3lib_div::trimExplode(',',$conf['linkOnlyIfNotEmpty']);
							if (count($linkOnlyIfNotEmpty)) {
								foreach ($linkOnlyIfNotEmpty as $checkField) {
									if ($item[$checkField] == '') {
										$generateLink = 0;
										break;
									}
								}
							}
						}

					} else {
						$generateLink = 0;
					}

					// create and wrap replacement
					// decide to preserve case of the displayed word in the content
					if ($item['force_preservecase']) {
						$replacement = ($item['force_preservecase'] == 1 ? "$1" : $item['short']);
					} else {
						$replacement = (intval($conf['preserveCase']) ? "$1" : $item['short']);
					}

					$replacement = trim($cObj->stdWrap($replacement, $conf[$element . '.']));
					$replacement = ' <' . $element . $lang . $title . '> ' . $replacement . ' </' . $element . '> ';

					if ($generateLink) {
						$replacement = ' ' . $cObj->typoLink($replacement, $conf['typolink.']) . ' ';
					}

					// set needle
					$needle = preg_quote($item['shortcut'] ? $item['shortcut'] : $item['short'],'/');

					// set needle modifiers
					$PCREmodifiers = $conf['patternModifiers'];
					if ($item['force_case']) {
						$caseSensitive = ($item['force_case'] == 1) ? 'i' : '';
					} else {
						$caseSensitive = $conf['caseSensitive'] ? '' : 'i';
					}

					// wrap needle regexp
					switch ($item['force_regexp']) {

						// word part
						case 1:
							$regExp = '(' . $needle . ')(?![^<>]*?[>])';
							break;

						// single word
						default:
							if ($generateLink) {
								// TODO get nested link recognition working,
								// feel free to contact us (info@artplan21.de)
								// if you know how to do better than this
								$regExp = '(?!<.*?)(?<=\s|[[:punct:]])(' . $needle . ')(?=\s|[[:punct:]])(?![^<>]*?'.'>)';

							} else {
								$regExp = '(?!<.*?)(?<=\s|[[:punct:]])(' . $needle . ')(?=\s|[[:punct:]])(?![^<>]*?>)';
							}
							break;
					}
					$regExp  = '/' . $regExp . '/' . $caseSensitive . $PCREmodifiers;

					// glossary area override for global replacement
					if ($item['force_global']) {
						$this->searchGlobal[] = $regExp;
						$this->replaceGlobal[] = $replacement;
					} else {
						$this->search[] = $regExp;
						$this->replace[] = $replacement;
					}
				}
			}

			// split content in head and body
			$dividePos = strpos($content, '<body');
			$head = substr($content, 0, $dividePos);
			$body = substr($content, strpos($content, '<body'));

			// text boundaries fix
			$body = str_replace('<', ' <', $body);
			$body = str_replace('>', '> ', $body);

			// prepare include-exclude parts
			$this->searchMarkers  = array(
				'<!--A21GLOSSARY_begin-->',
				'<!--A21GLOSSARY_end-->',
				'<!--A21GLOSSEX_begin-->',
				'<!--A21GLOSSEX_end-->'
			);
			$this->replaceMarkers = array(
				'<a21glossary>',
				'</a21glossary>',
				'<a21glossex>',
				'</a21glossex>'
			);


			// add predefined includes
			if ($conf['glossaryWHAT']=='ALL') {
				$this->searchMarkers[] = '<body>';
				$this->replaceMarkers[] = '<body><a21glossary>';

				$this->searchMarkers[] = '</body>';
				$this->replaceMarkers[] = '</a21glossary></body>';
			}

			if ($conf['glossaryWHAT']=='SEARCHTAGS' || intval($conf['includeSearchTags'])) {
				$this->searchMarkers[] = '<!--TYPO3SEARCH_begin-->';
				$this->replaceMarkers[] = '<!--TYPO3SEARCH_begin--><a21glossary>';

				$this->searchMarkers[] = '<!--TYPO3SEARCH_end-->';
				$this->replaceMarkers[] = '</a21glossary><!--TYPO3SEARCH_end-->';
			}

			if (strlen($conf['excludeTags'])) {

				$excludeTags = explode(',', strtolower($conf['excludeTags']));

				if (count($excludeTags)) {

					// also add uppercase version of each html element
					$excludeTags = array_merge($excludeTags,explode(',', strtoupper($conf['excludeTags'])));
					$excludeTags = array_unique($excludeTags);

					foreach ($excludeTags as $value) {
						$patternValue = preg_quote($value, '#');

						// b is for word boundary
						$this->searchMarkers2[] = '#<' . $patternValue . '\b#';
						$this->replaceMarkers2[] = '<a21glossex><' . $patternValue;

						$this->searchMarkers2[] = '#</' . $patternValue . '>#';
						$this->replaceMarkers2[] = '</' . $patternValue . '></a21glossex>';
					}
				}
			}



			// count entries
			$this->count['global'] = count($this->searchGlobal);
			$this->count['local'] = count($this->search);

			// replace global entries
			if ($this->count['global']) {
				$body = $this->replace($this->searchGlobal, $this->replaceGlobal, $body);
			}

			// replace local entries
			if ($this->count['local']) {

				// set splitmarkers
				$body = str_replace($this->searchMarkers, $this->replaceMarkers, $body);

				// replace local entries by recursive content splitting
				$this->parseObj = t3lib_div::makeInstance('t3lib_parsehtml');
				$body = $this->splitAndReplace($body);

				// final marker handling
				if ($conf['keepGlossaryMarkers']) {

					$body = str_replace(
						array('<a21glossary>', '</a21glossary>', '<a21glossex>', '</a21glossex>'),
						array('<!--A21GLOSSARY_begin-->', '<!--A21GLOSSARY_end-->', '<!--A21GLOSSEX_begin-->', '<!--A21GLOSSEX_end-->'),
						$body
					);

				} else {

					$body = str_replace(
						array('<a21glossary>', '</a21glossary>', '<a21glossex>', '</a21glossex>'),
						array('', '', '', ''),
						$body
					);
				}
			}

			// undo text boundaries fix
			$body = str_replace(' <', '<', $body);
			$body = str_replace('> ', '>', $body);

			return $head.$body;
		}
	}


	/**
	 * function that does the DB calls to fetch all glossary entries
	 *
	 * @param string $pidList idlists set by configuraion
	 * @return array glossary items
	 */
	protected function fetchGlossaryItems($pidList) {

		// -1 means: ignore pids
		if(!strlen(trim($pidList))) {
			$pidList = -1;
		}

		// fetch glossary items
		$pidList = t3lib_div::intExplode(',', $pidList);
		$languageUid = intval($GLOBALS['TSFE']->sys_language_uid);

		// manual ordering/grouping by pidlist
		foreach ($pidList as $pid) {

			$rows = $GLOBALS['TYPO3_DB']->exec_SELECTgetRows(
				'*',						// SELECT
				'tx_a21glossary_main',		// FROM
				'1=1'.						// WHERE
					($pid!=-1 ? ' AND pid=' . $pid : '').
					' AND tx_a21glossary_main.sys_language_uid IN (-1, ' . $languageUid . ')' .
					tslib_cObj::enableFields('tx_a21glossary_main'),
				'',							// GROUP BY
				'short,uid'					// ORDER BY
			);

			if (count($rows)) {
				foreach ($rows as $row) {
					$row['shortcut'] = trim($row['shortcut']);
					$row['short'] = trim($row['short']);
					$items[($row['shortcut'] ? $row['shortcut'] : $row['short'])] = $row;
				}
				$this->count['found'] += count($rows);
			}
		}

		$this->count['used'] = count($items);
		return $items;
	}


	/**
	 * wrapper function for preg_replace
	 *
	 * @param array $search the search expressions
	 * @param array $replace the replacement strings
	 * @param string $source
	 * @return the result count
	 */
	protected function replace($search = array(), $replace = array(), $content = '') {

		$content = preg_replace($search, $replace, $content, -1, $counter);
		$this->count['replaced'] += $counter;
		return $content;
	}


	/**
	 * splits the content recursivly into replaceable and unreplaceable parts
	 *
	 * @param string $body
	 * @param boolean $glossaryOn
	 * @param boolean $tagsExcluded
	 * @param integer $depth
	 * @return string
	 */
	protected function splitAndReplace($content, $glossaryOn = 0, $tagsExcluded = 0, $depth = 0) {

		// infinite failsafe
		if ($depth >= 15) {
			$this->depths[-1]++;
			return $content;
		}

		// split
		$contentSplit = $this->parseObj->splitIntoBlock('a21glossary,a21glossex', $content);

		// content is splittable
		if (count($contentSplit) > 1) {

			$result = '';
			foreach ($contentSplit as $contentSplitValue) {

				// replaceable part
				if ((substr($contentSplitValue, 0, 13) == '<a21glossary>')&&
					(substr($contentSplitValue, -14) == '</a21glossary>')) {

					$result.= '<a21glossary>' . $this->splitAndReplace(substr($contentSplitValue, 13, -14), 1, $tagsExcluded, $depth+1) . '</a21glossary>';

				// excluded part
				} elseif ((substr($contentSplitValue, 0, 12) == '<a21glossex>') &&
					(substr($contentSplitValue, -13) == '</a21glossex>')) {

					// change of rules: once excluded, nothing inside may be included again.
					// $result.= '<a21glossex>'.$this->splitAndReplace(substr($contentSplitValue,12,-13),0,$depth+1).'</a21glossex>';
					$result .= $contentSplitValue;

				// unknown part
				} else {

					$result.= $this->splitAndReplace($contentSplitValue, $glossaryOn, $tagsExcluded, $depth+1);
				}
			}
			return $result;

			// content seems not splittable
		} elseif ($glossaryOn) {

			// have you already passed this branch once?
			if ($tagsExcluded) {

				$this->depths[$depth]++;
				// content is not splittable - replace it, nonsplittable part and glossary enabled
				return $this->replace($this->search, $this->replace, $content);

			} else {

				// content maybe splittable if excludetags will be marked, so mark them and go on with recursion
				return $this->splitAndReplace(
					preg_replace(
						$this->searchMarkers2,
						$this->replaceMarkers2,
						$content
					),
					$glossaryOn,
					1,
					$depth + 1
				);
			}

			// dead end, nonsplittable part and glossary disabled
		} else {
			return $content;
		}
	}
}

?>